_posts/en/newsletters/2019-05-28-newsletter.md

```diff
@@ -0,0 +1,593 @@
+---
+title: 'Bitcoin Optech Newsletter #48'
+permalink: /en/newsletters/2019/05/28/
+name: 2019-05-28-newsletter
+type: newsletter
+layout: newsletter
+lang: en
+---
+This week's newsletter describes the new proposed
+`OP_CHECKOUTPUTSHASHVERIFY` opcode and covers continued discussion of
+Taproot.  Also included are our regular sections on bech32 sending
+support, top-voted Bitcoin StackExchange question and answers, and
+notable changes in popular Bitcoin infrastructure projects.
+
+{% comment %}<!-- include references.md below the fold but above any Jekyll/Liquid variables-->{% endcomment %}
+{% include references.md %}
+
+## Action items
+
+*None this week.*
+
+## News
+
+- **Proposed new opcode for transaction output commitments:** Jeremy
+  Rubin posted to the Bitcoin-Dev mailing list a proposal to soft fork
+  in an `OP_CHECKOUTPUTSHASHVERIFY` opcode that allows a Bitcoin address
+  to require the transaction spending it include a certain set of
+  outputs.  For details, please see the special section about the
+  proposal later in this newsletter.
+
+- **Continued discussion about bip-taproot and bip-tapscript:**
+  two comments this week from the Bitcoin-Dev mailing list discussion
+  seemed particularly noteworthy:
+
+    - *Final stack empty:* in legacy, segwit, and proposed
+      [bip-tapscript][] scripts, a script evaluates successfully if it
+      contains exactly[^legacy-cleanstack] one element that is *true*.
+      Russell O'Connor [raised][emptystack1] a point he's [raised
+      before][emptystack0] and requested that this opportunity be taken
+      to require tapscript only evaluate successfully if it ends with an
+      empty stack.  Pieter Wuille [replied][emptystack reply] that his
+      work on [miniscript][] (see [Newsletter #32][]) showed that, for
+      the subset of scripts miniscript will create, this change in
+      semantics will at most save 0.25 vbytes per tapscript.  Also,
+      although the change may also simplify development for anyone
+      writing scripts by hand, it's a bit risky as every development
+      guide to Script written to date teaches that scripts must
+      terminate with a *true* value on the stack.  Wuille summarized,
+      "so overall this feels like something with marginal costs, but
+      also at most marginal benefits."
+
+    - *Move the oddness byte:* Bitcoin public keys are graph points, so
```

Comment:
@jnewbery 

> Note that the proposal is to place the y-oddness bit in the leading byte of the control block, which has unused bits, so this doesn't increase the size of the witness

I phrased it that way because I don't understand what you and sipa were talking about.  You and he mention the control block, but IIUC, the control block isn't there for key-path spends.  For script-path spends where the control block is present, I don't think you need the Y bit as there's no signature verification being done---you can compare just the X coordinate of the output pubkey to just the X coordinate of ec_add(tweak, internal_key).

Reply:
For key path spends, you just always treat the key as if it were even (so imply an 0x02 prefix). If your secret key generated an odd key (0x03 prefix), just negate it to get the secret key for the even variant.

For script path spends you need the oddness in order to batch verification, because you need to verify Q = P+H(P,S)*G but if you're guessing Q's parity, then you might actually be plugging (-Q) = P+H(P,S)*G into the verifier, in which case you'll get a failure when you should be getting a success.